### ISSUE TITLE
Auto tuner takes too much time in serialized model

### ISSUE THREAD

**Issue by rrkarim** [pytorch/pytorch#24336]
## To Reproduce

I'm providing the trimmed version of `poseresnet` (`posenet.py`):
```
import torch
import torch.nn as nn

BN_MOMENTUM = 0.1

def conv3x3(in_planes, out_planes, stride=1):
    """3x3 convolution with padding"""
    return nn.Conv2d(in_planes, out_planes, kernel_size=3, stride=stride,
                     padding=1, bias=False)


class Bottleneck_nodown(nn.Module):
    expansion = 4
    def __init__(self, inplanes, planes, stride=1):
        super(Bottleneck_nodown, self).__init__()
        self.conv1 = nn.Conv2d(inplanes, planes, kernel_size=1, bias=False)
        self.bn1 = nn.BatchNorm2d(planes, momentum=BN_MOMENTUM)
        self.conv2 = nn.Conv2d(planes, planes, kernel_size=3, stride=stride,
                               padding=1, bias=False)
        self.bn2 = nn.BatchNorm2d(planes, momentum=BN_MOMENTUM)
        self.conv3 = nn.Conv2d(planes, planes * self.expansion, kernel_size=1,
                               bias=False)
        self.bn3 = nn.BatchNorm2d(planes * self.expansion,
                                  momentum=BN_MOMENTUM)
        self.relu = nn.ReLU(inplace=True)
        self.stride = stride

    def forward(self, x):
        residual = x

        out = self.conv1(x)
        out = self.bn1(out)
        out = self.relu(out)

        out = self.conv2(out)
        out = self.bn2(out)
        out = self.relu(out)

        out = self.conv3(out)
        out = self.bn3(out)

        out += residual
        out = self.relu(out)

        return out


class Bottleneck_down(nn.Module):
    expansion = 4

    def __init__(self, inplanes, planes, stride=1):
        super(Bottleneck_down, self).__init__()
        self.conv1 = nn.Conv2d(inplanes, planes, kernel_size=1, bias=False)
        self.bn1 = nn.BatchNorm2d(planes, momentum=BN_MOMENTUM)
        self.conv2 = nn.Conv2d(planes, planes, kernel_size=3, stride=stride,
                               padding=1, bias=False)
        self.bn2 = nn.BatchNorm2d(planes, momentum=BN_MOMENTUM)
        self.conv3 = nn.Conv2d(planes, planes * self.expansion, kernel_size=1,
                               bias=False)
        self.bn3 = nn.BatchNorm2d(planes * self.expansion,
                                  momentum=BN_MOMENTUM)
        self.relu = nn.ReLU(inplace=True)
        #self.downsample = downsample
        self.stride = stride
        self.downsample = nn.Sequential(
            nn.Conv2d(inplanes, planes * self.expansion,
                      kernel_size=1, stride=stride, bias=False),
            nn.BatchNorm2d(planes * self.expansion, momentum=BN_MOMENTUM),
        )

    def forward(self, x):
        residual = x

        out = self.conv1(x)
        out = self.bn1(out)
        out = self.relu(out)

        out = self.conv2(out)
        out = self.bn2(out)
        out = self.relu(out)

        out = self.conv3(out)
        out = self.bn3(out)

        residual = self.downsample(x)

        out += residual
        out = self.relu(out)

        return out


resnet_spec = {50: (Bottleneck_down, [3, 4, 6, 3]),
               101: (Bottleneck_down, [3, 4, 23, 3]),
               152: (Bottleneck_down, [3, 8, 36, 3])}


class PoseResNet_2(nn.Module):
    def __init__(self, block, layers, num_joints,
                 num_input_channels=3,
                 deconv_with_bias=False,
                 num_deconv_layers=3,
                 num_deconv_filters=(256, 256, 256),
                 num_deconv_kernels=(4, 4, 4),
                 final_conv_kernel=1,
                 ):
        super().__init__()

        self.num_joints = num_joints
        self.num_input_channels = num_input_channels
        self.inplanes = 64

        self.deconv_with_bias = deconv_with_bias
        self.num_deconv_layers, self.num_deconv_filters, self.num_deconv_kernels = num_deconv_layers, num_deconv_filters, num_deconv_kernels

        self.conv1 = nn.Conv2d(num_input_channels, 64, kernel_size=7, stride=2, padding=3,
                               bias=False)
        self.bn1 = nn.BatchNorm2d(64, momentum=BN_MOMENTUM)
        self.relu = nn.ReLU(inplace=True)
        self.maxpool = nn.MaxPool2d(kernel_size=3, stride=2, padding=1)
        self.layer1 = self._make_layer(block, 64, layers[0])
        self.layer2 = self._make_layer(block, 128, layers[1], stride=2)
        self.layer3 = self._make_layer(block, 256, layers[2], stride=2)


    def _make_layer(self, block, planes, blocks, stride=1):
        layers = []
        layers.append(block(self.inplanes, planes, stride))
        self.inplanes = planes * block.expansion
        for i in range(1, blocks):
            layers.append(block(self.inplanes, planes))

        return nn.Sequential(*layers)


    def forward(self, x):
        x = self.conv1(x)
        x = self.bn1(x)
        x = self.relu(x)
        x = self.maxpool(x)

        x = self.layer1(x)
        x = self.layer2(x)
        x = self.layer3(x)

        return x


def get_pose_net_2():
    block_class, layers = resnet_spec[152]
    model = PoseResNet_2(
        block_class, layers, 17,
    )
    return model
```

and the experiment file itself:

```
import torch
import torch.nn as nn
import poseresnet
import time

torch.backends.cudnn.benchmark=True

device = torch.device("cuda:0")

module = poseresnet.get_pose_net_2().to(device)
module = torch.jit.script(module)
print("serialized")

module(torch.randn((4, 3, 304, 304), device=device))
module(torch.randn((4, 3, 304, 304), device=device))

torch.cuda.synchronize()
start = time.time()

for i in range(400):
    module(torch.randn((4, 3, 304, 304), device=device))

torch.cuda.synchronize()
print(time.time() - start)
```

## Environment

 - PyTorch Version (e.g., 1.2.0):
 - OS (e.g., Linux):
 - GPU models and configuration: RTX 2080Ti

**Comment by zhangguanheng66** [comment]
I run the script and confirm the observation. Just to check. Do you see this performance issue since 1.2.0?

**Comment by rrkarim** [comment]
It's a new pytorch api with `torch.jit.script`. I was using decorators (already deprecated) to implement things.

**Comment by eellison** [comment]
Just to be clear: is it the compilation of the module that is slow or the first run of it ? 

cc: @Krovatkin

**Comment by rrkarim** [comment]
The compilation and the first run. If I understand correctrly how the new recursive serialization works, it doesn't need the first run to be compiled. In the provided example, I agree, the compilation is prettty slow (which is again the issue). Just remove the third layer, then the compilation will take a reasonable time and you will see how slow the first run is.

**Comment by suo** [comment]
@rrkarim yeah, right now a large model like yours will be quite slow to deserialize/compile, because we are serializing tons of redundant information. We are actively working on this exact issue, and it should be drastically improved soon.

**Comment by rrkarim** [comment]
@suo it is not the largest model I'm using now. The largest will include two additional layers, so it becomes super irrational to use it in production. But my concern is not the compilation speed but rather the first forward call. It takes unreasonably too much time, can you explain why is this the case after the model have been compiled?

**Comment by suo** [comment]
@rrkarim The compiler is a JIT compiler, so it specializes the graph based on the properties of the input and optimizes using that information. So the first time, the extra time you are seeing is the JIT compilation.

Some amount of warmup latency is expected; production services at Facebook generally will feed a representative set of inputs so that when the model is marked available the latency will be more predictable.

That said, 3-4 minutes is waaay more than an acceptable time to warm up. I'll dig into the repro script and see what's going on.

**Comment by suo** [comment]
this PR substantially improves the optimization speed (around 7x): https://github.com/pytorch/pytorch/pull/24827

**Comment by rrkarim** [comment]
Thanks, @suo. But if warm up step takes some time to jit compile the model, why not introducing the argument to the `torch.jit.script` which takes `tensor` with appropriate shape. (maybe we pass the tensor now but I don't see that in docs.) What is the reason of having the model before jit compilation?

**Comment by suo** [comment]
@rrkarim, it's a reasonable approach, but there are a few problems with it today.
- We don't serialize the input specializations, so they would not survive past a save/load cycle
- For models that expect variable input shapes, it only pushes the problem out one step further, since there will be a new optimization kicked off whenever we get a cache miss.

I think eventually we will need to support a totally ahead-of-time compilation mode with an error if you get unexpected shapes, because that's required for certain deployment environments (like mobile). But that's quite far away.

**Comment by rrkarim** [comment]
I see, its really hard to find any relevant information about the new specification, so thank you.

**Comment by suo** [comment]
What do you mean the new specification? We'd be happy to write some more docs if you have specific areas of confusion or places where you can't find good info

**Comment by rrkarim** [comment]
Well, I didn't know the whole process of how the serialization is done. Since there is the jit compilation on the first run, it is pretty self evident why there is the latency. 

I was confused by the documentation which states: `Scripting an nn.Module by default will compile the forward method and recursively compile any methods, submodules`. This made me believe that there is no need for the compilation after the wrapping. But since the input specializations are not stored (and how they can be, since we are not providing any), the warmup is going to take more time.

**Comment by janeyx99** [comment]
TorchScript has been deprecated in favor of torch.export as of 2.11. https://pytorch.org/blog/pytorch-2-11-release-blog/ Closing this issue as we are not further developing on it--please ping if you believe this issue was erroneously closed.
